src/workerd/jsg/modules-new.c++

```diff
@@ -0,0 +1,1401 @@
+  if (!allowsFallback()) {
+    KJ_REQUIRE(bundle->type() != ModuleBundle::Type::FALLBACK,
+               "Fallback bundle types are not allowed for this registry");
+  }
+  bundles_[static_cast<int>(bundle->type())].add(kj::mv(bundle));
+  return *this;
+}
+
+kj::Own<ModuleRegistry> ModuleRegistry::Builder::finish() {
+  return kj::heap<ModuleRegistry>(this);
+}
+
+ModuleRegistry::ModuleRegistry(ModuleRegistry::Builder* builder)
+    : observer(builder->observer),
+      maybeParent(builder->maybeParent) {
+  bundles_[kBundle] = builder->bundles_[kBundle].releaseAsArray();
+  bundles_[kBuiltin] = builder->bundles_[kBuiltin].releaseAsArray();
+  bundles_[kBuiltinOnly] = builder->bundles_[kBuiltinOnly].releaseAsArray();
+  bundles_[kFallback] = builder->bundles_[kFallback].releaseAsArray();
+}
+
+void ModuleRegistry::attachToIsolate(
+    Lock& js,
+    const CompilationObserver& observer) {
+  // We use new here because it is required for the v8 integration.
+  // Responsibility for cleaning up the type is passed to v8 when
+  // the context is cleaned up.
+  new IsolateModuleRegistry(js, *this, observer);
+}
+
+kj::Maybe<Module&> ModuleRegistry::resolve(const ResolveContext& context) {
+  constexpr auto tryFind = [](
+      const ResolveContext& context,
+      kj::ArrayPtr<kj::Own<ModuleBundle>> bundles) -> kj::Maybe<Module&> {
+    for (auto& bundle : bundles) {
+      KJ_IF_SOME(found, bundle->resolve(context)) {
+        return found;
+      }
+    }
+    return kj::none;
+  };
+
+  // If the embedder supports it, collect metrics on what modules were resolved.
+  auto metrics = observer.onResolveModule(context.specifier, context.type, context.source);
+
+  switch (context.type) {
+    case ResolveContext::Type::BUNDLE: {
+      // For bundle resolution, we only use Bundle, Builtin, and Fallback bundles,
+      // in that order.
+      KJ_IF_SOME(found, tryFind(context, bundles_[kBundle])) {
+        metrics->found();
+        return found;
+      }
+      KJ_IF_SOME(found, tryFind(context, bundles_[kBuiltin])) {
+        metrics->found();
+        return found;
+      }
+      KJ_IF_SOME(found, tryFind(context, bundles_[kFallback])) {
+        metrics->found();
+        return found;
+      }
+
+      KJ_IF_SOME(parent, maybeParent) {
+        return parent.resolve(context);
+      }
+
+      metrics->notFound();
+      return kj::none;
+     }
+    case ResolveContext::Type::BUILTIN: {
+      // For built-in resolution, we only use builtin and builtin-only bundles.
+      KJ_IF_SOME(found, tryFind(context, bundles_[kBuiltin])) {
+        metrics->found();
+        return found;
+      }
+      KJ_IF_SOME(found, tryFind(context, bundles_[kBuiltinOnly])) {
+        metrics->found();
+        return found;
+      }
+
+      KJ_IF_SOME(parent, maybeParent) {
+        return parent.resolve(context);
+      }
+
+      metrics->notFound();
+      return kj::none;
+    }
+    case ResolveContext::Type::BUILTIN_ONLY: {
```

Comment:
How frowned upon are case fallthroughs nowadays? The steps for BUILTIN_ONLY are a suffix of those for BUILTIN, so we could merge them if we find a fallthrough acceptable and add a comment stating that it's intentional.

Reply:
For now, I think I'd rather keep these separate even if there's a bit of duplication. We can come back and de-dup in a later pass.